### PR TITLE
Cleanup and de-dupe TaintIssuesSynchronizerTests

### PR DESCRIPTION
**Pull request by duncanp-sonar**

### CODE REVIEW

**Comment by duncanp-sonar** [comment]
@ugras-ergun-sonarsource these changes are going to harder to review in GitHub. I changed the order so logically similar tests are grouped together (e.g. tests that expect issues to be returned). I also deleted some redundant tests and merged a few others. Unfortunately, the GitHub UI doesn't do a good job of displaying these kind of changes.

File: src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs [hunk 1]
```diff
@@ -61,38 +61,60 @@ public void MefCtor_CheckIsExported()
         }
 
         [TestMethod]
-        public void SynchronizeWithServer_FailureToCheckServerConnection_CriticalException_ExceptionThrown()
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs:64]: Redundant test - deleted. There were multiple tests that checked critical/non-critical errors. I've kept one of each.

File: src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs [hunk 2]
```diff
@@ -61,38 +61,60 @@ public void MefCtor_CheckIsExported()
...
                 bindingConfig: BindingConfig_Connected,
-                sonarService: sonarQubeServer.Object);
+                logger: logger,
+                sonarService: sonarServer.Object);
 
             Func<Task> act = testSubject.SynchronizeWithServer;
+            await act.Should().NotThrowAsync();
 
-            act.Should().Throw<StackOverflowException>();
+            logger.AssertPartialOutputStringExists("this is a test");
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs:80]: I added the logging check to this test. Previously, it was in a separate test, which I've deleted.

File: src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs [hunk 3]
```diff
@@ -61,38 +61,60 @@ public void MefCtor_CheckIsExported()
...
         }
 
         [TestMethod]
-        public void SynchronizeWithServer_FailureToCheckServerConnection_NonCriticalException_ExceptionIsCaught()
+        public async Task SynchronizeWithServer_NonCriticalException_UIContextAndStoreCleared()
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs:84]: There are quite a few tests that check that `_UIContextAndStoreCleared` in various scenarios (e.g. non-critical exception, standalone mode, no issues returned).

Previously, some tests checked the UIContext, some checked StoreCleared, some checked both.
Now, all tests check both.

File: src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs [hunk 4]
```diff
@@ -201,153 +267,43 @@ public async Task SynchronizeWithServer_SupportedServer_IssuesFetched(ServerType
...
-                sonarService: sonarService.Object);
-
-            await testSubject.SynchronizeWithServer();
-
-            converter.Invocations.Should().HaveCount(0);
-            CheckStoreIsCleared(taintStore);
-        }
-
-        [TestMethod]
-        public async Task SynchronizeWithServer_FailureToSync_StoreCleared()
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs:245]: Redundant exception test.

File: src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs [hunk 5]
```diff
@@ -201,153 +267,43 @@ public async Task SynchronizeWithServer_SupportedServer_IssuesFetched(ServerType
...
-                taintStore: taintStore.Object,
-                sonarService: sonarServerMock.Object);
-
-            await testSubject.SynchronizeWithServer();
-
-            CheckStoreIsCleared(taintStore);
-        }
-
-        [TestMethod]
-        public async Task SynchronizeWithServer_NonCriticalException_ExceptionCaughtAndLogged()
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs:262]: Redundant exception test.

File: src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs [hunk 6]
```diff
@@ -201,153 +267,43 @@ public async Task SynchronizeWithServer_SupportedServer_IssuesFetched(ServerType
...
-                sonarService: sonarServerMock.Object);
-
-            Func<Task> act = testSubject.SynchronizeWithServer;
-            await act.Should().NotThrowAsync();
-
-            logger.AssertPartialOutputStringExists("this is a test");
-        }
-
-        [TestMethod]
-        public async Task SynchronizeWithServer_CriticalException_ExceptionNotCaught()
```
Comment (duncanp-sonar) [inline src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs:282]: Redundant exception test.

File: src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs [hunk 7]
```diff
@@ -60,38 +61,60 @@ public void MefCtor_CheckIsExported()
         }
 
         [TestMethod]
-        public void SynchronizeWithServer_FailureToCheckServerConnection_CriticalException_ExceptionThrown()
+        public async Task SynchronizeWithServer_NonCriticalException_ExceptionCaughtAndLogged()
```
Comment (ugras-ergun-sonarsource) [inline src/IssueViz.Security.UnitTests/Taint/TaintIssuesSynchronizerTests.cs:64]: it seems this test and the one below is mostly the same they only assert different things. Maybe they can be merged into one.

**Review by ugras-ergun-sonarsource — APPROVED** [review]
LGTM. Only one minor simplification we could consider.

**Pull request merged**